pkg/grpc/server.go

```diff
@@ -0,0 +1,249 @@
+		for i := 0; i < len(resp.Resources); i++ {
+			data, err := proto.Marshal(resp.Resources[i])
+			if err != nil {
+				// TODO(kuat) logging
+				continue
+			}
+			resources[i] = &any.Any{
+				TypeUrl: typeURL,
+				Value:   data,
+			}
+		}
+
+		out := &api.DiscoveryResponse{
+			VersionInfo: resp.Version,
+			Resources:   resources,
+			Canary:      resp.Canary,
+			TypeUrl:     typeURL,
+			Nonce:       strconv.FormatInt(nonce, 10),
+		}
+		nonce = nonce + 1
+		return stream.Send(out)
+	}
+
+	var values watches
+	defer func() {
+		values.cancel()
+	}()
+
+	for {
+		select {
+		// config watcher can send the requested resources types in any order
+		case resp, closed := <-values.endpoints.Value:
+			if closed {
+				return status.Errorf(codes.Unavailable, "endpoints watch failed")
+			}
+			if err := send(resp, EndpointType); err != nil {
+				return err
+			}
+		case resp, closed := <-values.clusters.Value:
+			if closed {
+				return status.Errorf(codes.Unavailable, "clusters watch failed")
+			}
+			if err := send(resp, ClusterType); err != nil {
+				return err
+			}
+		case resp, closed := <-values.routes.Value:
+			if closed {
+				return status.Errorf(codes.Unavailable, "routes watch failed")
+			}
+			if err := send(resp, RouteType); err != nil {
+				return err
+			}
+		case resp, closed := <-values.listeners.Value:
+			if closed {
+				return status.Errorf(codes.Unavailable, "listeners watch failed")
+			}
+			if err := send(resp, ListenerType); err != nil {
+				return err
+			}
+		case req, closed := <-reqCh:
+			switch {
+			case closed:
+				// input stream ended or errored out
+				return nil
+			case req.GetResponseNonce() != "" && req.GetResponseNonce() != strconv.FormatInt(nonce, 10):
+				// ignore stale non-empty nonces
+			default:
+				// proxy requests a new resource
+				// proxy can NACKs by sending an older version for the same resource type
+				typeURL := implicitTypeURL
+				if req.TypeUrl != "" {
+					typeURL = req.TypeUrl
+				}
+				// cancel existing watches to (re-)request a newer version
+				switch typeURL {
+				case EndpointType:
+					values.endpoints.Cancel()
+					values.endpoints = s.config.WatchEndpoints(req.GetNode(), req.GetVersionInfo(), req.GetResourceNames())
+				case ClusterType:
+					values.clusters.Cancel()
+					values.clusters = s.config.WatchClusters(req.GetNode(), req.GetVersionInfo(), req.GetResourceNames())
+				case RouteType:
+					values.routes.Cancel()
+					values.routes = s.config.WatchRoutes(req.GetNode(), req.GetVersionInfo(), req.GetResourceNames())
+				case ListenerType:
+					values.listeners.Cancel()
+					values.listeners = s.config.WatchListeners(req.GetNode(), req.GetVersionInfo(), req.GetResourceNames())
+				}
+			}
+		}
+	}
+}
+
+func (s *server) handler(stream stream, implicitTypeURL string) error {
+	// a channel for receiving incoming requests
+	reqCh := make(chan *api.DiscoveryRequest, 0)
+	reqStop := int32(0)
```

Comment:
What does this do? Looks funny

Reply:
It is funny. The problem here is that `recv` can block past the exit of the outer function since we can't stop the stream in grpc-go.

Reply:
@kyessenov can't you use the stream Context to accomplish the same thing? IIRC it will propagated if the client closes/cancel the stream but also if the underlying transport terminates (due to missed keepalive for example)?

(The same Context could also be used to derive child context for `watches`)

Reply:
Are you suggesting to poll `context.Err()` after unblocking from `Recv()`?